api/camunda/camunda-openapi.yaml

```diff
@@ -2748,7 +2923,150 @@ components:
           description: The name of the resource from which this decision requirements was parsed.
         tenantId:
           type: string
-          description: The tenant id of the decision requirements.
+          description: The tenant ID of the decision requirements.
+    EvaluateDecisionRequest:
+      type: object
+      properties:
+        decisionKey:
+          description: |
+            The unique key identifying the decision to be evaluated.
+            Cannot be used together with decisionId.
+          type: integer
+          format: int64
+        decisionId:
+          description: |
+            The ID of the decision to be evaluated.
+            Cannot be used together with decisionKey. When using the decision ID, the latest
```

Comment:
@tmetzke: @jwulf brought up in Slack that a similar endpoint, in which the request can contain _either_ an ID or a key but not both, was resulting in confusing docs, because the example request included both an ID and a key. 

There are a couple things we should do in our spec to account for these types of situations. I'm bringing them up here to start a conversation, though I will merge this PR as-is, and we can handle any resolutions outside of this. I'm interested to know how you feel about these suggestions, who you think should own a PR to update this specific spec, and how you think we can communicate this practice to the engineers developing specs.

## The proposed changes

### 1. Spec union types more intentionally

The `EvaluateDecisionRequest` can be spec'd such that the schema eliminates the possibility of including both an ID and a key. I _think_ the correct way to do this is with [the `oneOf` union](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/), though I am no OpenAPI expert and if others know of a better mechanism, I defer to them 😅. Here's an example of how we can write this request object so that it can only include an ID _or_ a key:

   ```yaml
    EvaluateDecisionRequest:
      type: object
      oneOf:
        - $ref: '#/components/schemas/EvaluateDecisionRequestKey'
        - $ref: '#/components/schemas/EvaluateDecisionRequestId'
    EvaluateDecisionRequestKey:
      type: object
      allOf:
        - $ref: '#/components/schemas/EvaluateDecisionRequestBase'
      properties: 
        decisionKey:
          description: |
            The unique key identifying the decision to be evaluated.
            Cannot be used together with decisionId.
          type: integer
          format: int64
    EvaluateDecisionRequestId:
      type: object
      allOf:
        - $ref: '#/components/schemas/EvaluateDecisionRequestBase'
      properties: 
        decisionId:
          description: |
            The ID of the decision to be evaluated.
            Cannot be used together with decisionKey. When using the decision ID, the latest
              deployed version of the decision is used.
          type: string
    EvaluateDecisionRequestBase:
      type: object
      properties: 
        variables:
          description: The message variables as JSON document.
          additionalProperties: true
          type: object
        tenantId:
          description: The tenant ID of the decision.
          type: string
   ```

   Here's how that schema gets rendered by Swagger: 

   <img width="807" alt="image" src="https://github.com/user-attachments/assets/97f6c105-510e-4b48-8d15-ae39074ed76a">

   And here's how it would get rendered by our docs: 

   <img width="733" alt="image" src="https://github.com/user-attachments/assets/def855b4-388d-426a-9b28-7e0265b1cb88">

### 2. Add example requests

As it currently stands, the example request in the docs is improperly formed -- it contains both an ID _and_ a key. Addressing item 1 above would resolve this, in that the docs will generate a properly formed request given the union type described above.

But we can make this even better, by providing example requests for the different possibilities. Here's an example modification to the endpoint request definition, so that two examples are provided: 

```yaml
  /decisions/evaluation:
    post:
      tags:
        - Decision
      summary: Evaluate decision
      description: |
        Evaluates a decision.
        You specify the decision to evaluate either by using its unique key (as returned by
        DeployResource), or using the decision ID. When using the decision ID, the latest deployed
        version of the decision is used.
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/EvaluateDecisionRequest"
            examples:
              "Decision Key":
                summary: "Identify the decision by key"
                value:
                  decisionKey: 123
                  variables: { }
                  tenantId: 987
              "Decision ID":
                summary: "Identify the decision by ID"
                value:
                  decisionId: "1234-5678"
                  variables: { }
                  tenantId: 987
...
```

Here's how the examples appear in Swagger UI: 

<img width="817" alt="image" src="https://github.com/user-attachments/assets/852cf6fb-dc1c-4f31-a61c-4b1a20af9d30">

And here's how they'd appear in our docs: 

<img width="524" alt="image" src="https://github.com/user-attachments/assets/69976d17-704e-40f1-9ba2-80a03973ecef">

Reply:
Both of these - strict grammar and meaningful representative example data - will make a huge difference.

Reply:
Thanks for bringing this up and great proposal, @pepopowitz 👍 

One aspect of the `oneOf` mechanism that OpenAPI provides is that it poses an issue with the Java models we generate from it. There are multiple issues with this polymorphism mechanism in the code generators and not only for Java, if I remember correctly. The generators struggle to create code that works in an API setup where data needs to be (de-)serialized, leading to custom post-processing effort required by generator users (us, in this case) to make it work. I would like to avoid introducing this risk for this and other endpoints, especially now, since the code freeze for 8.6 is next Monday 😅 
Thus, I propose to go the pragmatic way of adding proper examples to the spec and leaving the endpoint structurally as it is.

Let me know if that makes sense or if I missed anything.

Reply:
> Does anyone think it would cause recurring difficulties if we left the Java-friendly definition alone in the camunda repo, but swapped in the docs-friendly definition in the camunda-docs repo? If not, I can PR this. In fact, I think I'll open the PR regardless, and we can have more conversation there about the risks/rewards it introduces.

Thanks for opening this conversation @pepopowitz. Can you bring this to ask-developer-experience so we can chat more on Slack?

Reply:
Temporary example requests are added on the docs side at https://github.com/camunda/camunda-docs/pull/4352.